upup/pkg/fi/cloudup/awstasks/dnszoneassociation.go

```diff
@@ -0,0 +1,119 @@
+package awstasks
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/golang/glog"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/awsup"
+	"k8s.io/kops/upup/pkg/fi/cloudup/terraform"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"k8s.io/kubernetes/pkg/util/validation/field"
+)
+
+//go:generate fitask -type=DNSZoneAssociation
+type DNSZoneAssociation struct {
+	Name    *string
+	DNSZone *DNSZone
+	VPC     *VPC
+}
+
+var _ fi.CompareWithID = &AutoscalingGroup{}
+
+func (e *DNSZoneAssociation) CompareWithID() *string {
+	return e.Name
+}
+
+func (e *DNSZoneAssociation) Find(c *fi.Context) (*DNSZoneAssociation, error) {
+	cloud := c.Cloud.(awsup.AWSCloud)
+
+	dnsZoneID := e.DNSZone.ID
+	vpcID := e.VPC.ID
+
+	if dnsZoneID == nil || vpcID == nil {
+		return nil, nil
+	}
+
+	request := &route53.GetHostedZoneInput{
+		Id: dnsZoneID,
+	}
+
+	response, err := cloud.Route53().GetHostedZone(request)
+	if err != nil {
+		return nil, fmt.Errorf("error listing HostedZone %q: %v", dnsZoneID, err)
+	}
+	if response == nil {
+		return nil, nil
+	}
+
+	for _, v := range response.VPCs {
+		if aws.StringValue(v.VPCId) != *vpcID {
+			continue
+		}
+		if aws.StringValue(v.VPCRegion) != cloud.Region() {
+			// This seems incredibly unlikely
+			glog.Warningf("Found matching VPC ID, but region mismatch: %q vs %q", aws.StringValue(v.VPCRegion), cloud.Region())
+			continue
+		}
+		actual := &DNSZoneAssociation{
+			DNSZone: e.DNSZone,
+			VPC:     e.VPC,
+		}
+		glog.V(2).Infof("found matching DNSZoneAssociation")
+		return actual, nil
+	}
+
+	return nil, nil
+}
+
+func (e *DNSZoneAssociation) Run(c *fi.Context) error {
+	return fi.DefaultDeltaRunMethod(e, c)
+}
+
+func (s *DNSZoneAssociation) CheckChanges(a, e, changes *DNSZoneAssociation) error {
```

Comment:
a = actual, e = expected.  It's what we're dealing with everywhere here, so it becomes a pattern.